sql/parser: correctly escape non-printable bytes

madelynnblue opened the pull request:
Fix the output of the 0 byte, which was producing `b'\0'`, which isn't a
valid byte string. Instead output `b'\x00'`.

Change the other non-printable bytes to use escape characters instead of
preserving the byte. Although within cockroach itself, those bytes are
parsed correctly, if they are ever persisted to a file and then read in
via `cockroach sql`, an error would occur because the readline package
interprets those bytes as if a user were typing them, and some of them
are readline commands. Escaping the bytes makes them easy to understand
in a file and doesn't break the readline input.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7294)

<!-- Reviewable:end -->

knz commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/encode.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQ0IpvT6sJWYp0rC25:-KKQ0IpvT6sJWYp0rC26:1083214254) ([raw file](https://github.com/cockroachdb/cockroach/blob/83665e561b16d4c726a00cb4a52475aea92868c1/sql/parser/encode.go#L103)):*
> ```Go
> 			buf.WriteByte(encodedChar)
> 			start = i + 1
> 		} else if ch >= 0x79 || ch <= 0x1F {
> ```

This needs to be `ch >= 0x7F`. Also usually people say `ch < 0x20` as 0x20 is the well-known first printable ASCII character.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294)*
<!-- Sent from Reviewable.io -->

tamird commented:
Do we need a CLI test?

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/encode.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQ0IpvT6sJWYp0rC25:-KKQ0vqzu4iDgPkojRvP:1869606791) ([raw file](https://github.com/cockroachdb/cockroach/blob/83665e561b16d4c726a00cb4a52475aea92868c1/sql/parser/encode.go#L103)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> This needs to be `ch >= 0x7F`. Also usually people say `ch < 0x20` as 0x20 is the well-known first printable ASCII character.
</details>
How about a comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-:-KKQ1FbOYosYkNvcG-5G:1898517670)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/encode.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQ0IpvT6sJWYp0rC25:-KKQ3612s1QjWIpoG45T:592770326) ([raw file](https://github.com/cockroachdb/cockroach/blob/83665e561b16d4c726a00cb4a52475aea92868c1/sql/parser/encode.go#L103)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> How about a comment?
</details>
Comment works too. But 0x7F instead of 0x79 is rather important. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/parser/parse_test.go, line 1116 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQ5t-AeX4qC44cS1hz:-KKQ5t-Bvu85H2hQPBO_:1866723234) ([raw file](https://github.com/cockroachdb/cockroach/blob/83665e561b16d4c726a00cb4a52475aea92868c1/sql/parser/parse_test.go#L1116)):*
> ```Go
> func TestEncodeSQLBytes(t *testing.T) {
> 	for i := 0; i < 256; i++ {
> 		s := string([]byte{byte(i)})
> ```

I just saw another test fail in the CI results for this PR. I think you need to test multi-byte combinations here as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294)*
<!-- Sent from Reviewable.io -->

mjibson commented:
We probably need a CLI test, yes. But it's difficult because it involves making a test that goes through the readline code (we don't have any tests that do this yet). I mostly just don't want to. Maybe it's not that hard...

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/parser/encode.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQ0IpvT6sJWYp0rC25:-KKQ3FHUbO1F9vfcmPXO:-2055665624) ([raw file](https://github.com/cockroachdb/cockroach/blob/83665e561b16d4c726a00cb4a52475aea92868c1/sql/parser/encode.go#L103)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> How about a comment?
</details>
Oops. 0x80 - 1 != 0x79.

---

*[sql/parser/parse_test.go, line 1116 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQ5t-AeX4qC44cS1hz:-KKQ82_ZcFnfQqH92QDQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/83665e561b16d4c726a00cb4a52475aea92868c1/sql/parser/parse_test.go#L1116)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I just saw another test fail in the CI results for this PR. I think you need to test multi-byte combinations here as well.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-:-KKQ83sApFdyJsjz9II8:-1905469775)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/parser/encode.go, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQ0IpvT6sJWYp0rC25:-KKQAtzRm9b_cDPsv93M:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/83665e561b16d4c726a00cb4a52475aea92868c1/sql/parser/encode.go#L103)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Oops. 0x80 - 1 != 0x79.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/parse_test.go, line 1134 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQBIXcq5HvauKfWCbb:-KKQBIXcq5HvauKfWCbc:-1827353987) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c8025603c97e17eaf485d4e25696545a9933d82/sql/parser/parse_test.go#L1134)):*
> ```Go
> 	// Test all multi-bytes combinations.
> 	for i := 0; i < 256; i++ {
> 		for j := 0; j < 256; j++ {
> ```

I'm worried this will take a while to run.
I was more thinking about investigating what went wrong in the CI results for the previous version and make a test just for that case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql/parser: correctly escape non-printable bytes

Fix the output of the 0 byte, which was producing `b'\0'`, which isn't a
valid byte string. Instead output `b'\x00'`.

Change the other non-printable bytes to use escape characters instead of
preserving the byte. Although within cockroach itself, those bytes are
parsed correctly, if they are ever persisted to a file and then read in
via `cockroach sql`, an error would occur because the readline package
interprets those bytes as if a user were typing them, and some of them
are readline commands. Escaping the bytes makes them easy to understand
in a file and doesn't break the readline input.

mjibson commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/parse_test.go, line 1134 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-KKQBIXcq5HvauKfWCbb:-KKQCZlOpiKurCa8-Uvk:1385635905) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c8025603c97e17eaf485d4e25696545a9933d82/sql/parser/parse_test.go#L1134)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'm worried this will take a while to run.
> I was more thinking about investigating what went wrong in the CI results for the previous version and make a test just for that case.
</details>
The CI error was due to the `0x79` bug. I've removed the multi-byte cases. I think this is fine now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-:-KKQE6ZM15aqeKC_UjC5:1417088721)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Looking at it closer, I'm voting against the CLI test for now. It doesn't appear to be written in a way that's convenient to test, and it wasn't being tested before anyway.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294#-:-KKQE_s4k9wBHtKKe7LK:-1686727536)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7294)*
<!-- Sent from Reviewable.io -->
